# [Python for Data Journalism](http://journalismcourses.org/course/view.php?id=52 "Knight Center for Journalism in the Americas")—California Propositions Analysis
Summer 2017

*see also [First Python Notebook](http://www.firstpythonnotebook.org/index.html)*

In [2]:
import pandas as pd

Read in proposition and committee data, from [California Civic Data Coalition](https://www.californiacivicdata.org/).

In [3]:
props = pd.read_csv("http://www.firstpythonnotebook.org/_static/committees.csv")

In [4]:
props.head(3)

ocd_prop_id  calaccess_prop_id  \
0  ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
1  ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   
2  ocd-contest/b51dc64d-3562-4913-a190-69f5088c22a6            1376258   

   ccdc_prop_id                                          prop_name  \
0            70  PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...   
1            70  PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...   
2            70  PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...   

   ccdc_committee_id  calaccess_committee_id  \
0                382                 1374469   
1                383                 1220380   
2                384                 1282321   

                                      committee_name committee_position  
0  YES ON PROPOSITION 51 - CALIFORNIANS FOR QUALI...            SUPPORT  
1  COMMUNITY COLLEGE FACILITY COALITION ISSUES CO...            SUPPORT  
2  TORLAKSON'S INVEST IN CALIFORNIA A BALLOT MEAS...            SUPPORT

In [5]:
props.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102 entries, 0 to 101
Data columns (total 8 columns):
ocd_prop_id               102 non-null object
calaccess_prop_id         102 non-null int64
ccdc_prop_id              102 non-null int64
prop_name                 102 non-null object
ccdc_committee_id         102 non-null int64
calaccess_committee_id    102 non-null int64
committee_name            102 non-null object
committee_position        102 non-null object
dtypes: int64(4), object(4)
memory usage: 6.5+ KB


Read in contribution data for these committees.

In [6]:
contribs = pd.read_csv("http://www.firstpythonnotebook.org/_static/contributions.csv")

In [7]:
contribs.head(3)

calaccess_committee_id                              committee_name  \
0                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
1                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   
2                 1386560  ADULT USE CAMPAIGN FOR PROPOSITION 64; THE   

   calaccess_filing_id date_received contributor_lastname  \
0              2083796    2016-09-18              BERGMAN   
1              2083796    2016-09-18                KAHLE   
2              2083796    2016-07-15             MCDEVITT   

  contributor_firstname contributor_city contributor_state contributor_zip  \
0              GRETCHEN    SPRING VALLEY                CA           91978   
1                 MYRNA        SAN DIEGO                CA           92109   
2                   LEO        ESCONDIDO                CA           92025   

       contributor_employer contributor_occupation  \
0                A NEW PATH     EXECUTIVE DIRECTOR   
1  NATIONAL SCHOOL DISTRICT                TEACHER   
2             LIFE IONIZERS    SEO/CONTENT MANAGER   

   contributor_is_self_employed  amount  
0                         False    84.0  
1                         False    35.0  
2                         False   198.0

In [8]:
contribs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56379 entries, 0 to 56378
Data columns (total 13 columns):
calaccess_committee_id          56379 non-null int64
committee_name                  56379 non-null object
calaccess_filing_id             56379 non-null int64
date_received                   56379 non-null object
contributor_lastname            56375 non-null object
contributor_firstname           53411 non-null object
contributor_city                56369 non-null object
contributor_state               56363 non-null object
contributor_zip                 56366 non-null object
contributor_employer            48572 non-null object
contributor_occupation          53273 non-null object
contributor_is_self_employed    56379 non-null bool
amount                          56379 non-null float64
dtypes: bool(1), float64(1), int64(2), object(9)
memory usage: 5.2+ MB


Show proposition names and committee counts, pretty-printing with `reset_index()`.

In [9]:
props.prop_name.value_counts().reset_index()

index  prop_name
0   PROPOSITION 057 - CRIMINAL SENTENCES. JUVENILE...         13
1   PROPOSITION 056 - CIGARETTE TAX TO FUND HEALTH...         12
2   PROPOSITION 064- MARIJUANA LEGALIZATION. INITI...         11
3   PROPOSITION 066- DEATH PENALTY. PROCEDURES. IN...          9
4   PROPOSITION 055 - TAX EXTENSION TO FUND EDUCAT...          8
5   PROPOSITION 067- REFERENDUM TO OVERTURN BAN ON...          7
6   PROPOSITION 062- DEATH PENALTY. INITIATIVE STA...          7
7   PROPOSITION 059- SB 254 (CHAPTER 20, STATUTES ...          6
8   PROPOSITION 053 - REVENUE BONDS. STATEWIDE VOT...          4
9   PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...          4
10  PROPOSITION 054 - LEGISLATURE. LEGISLATION AND...          4
11  PROPOSITION 051 - SCHOOL BONDS. FUNDING FOR K-...          4
12  PROPOSITION 063- FIREARMS. AMMUNITION SALES. I...          4
13  PROPOSITION 061- STATE PRESCRIPTION DRUG PURCH...          3
14  PROPOSITION 052 - STATE FEES ON HOSPITALS. FED...          3
15  PROPOSITION 060- ADULT FILMS. CONDOMS. HEALTH ...          2
16  PROPOSITION 065- CARRY-OUT BAGS. CHARGES. INIT...          1

A curious statistic.

In [10]:
contribs.amount.median()

287.08

Show *complete* committee names and contribution counts.

In [12]:
contribs.committee_name.value_counts().head()

CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIATION TRUTH IN AMERICAN GOVERNMENT FUND; NO ON PROPOSITION 62, YES ON PROPOSITION 66 (NON-PROFIT 501 (C) 5)    28617
NEWSOM BALLOT MEASURE COMMITTEE; SAFETY FOR ALL, YES ON PROP. 63,                                                                                            5457
YES ON 62, NO ON 66. REPLACE THE COSTLY, FAILED DEATH PENALTY SYSTEM. SPONSORED BY TAXPAYERS FOR SENTENCING REFORM                                           4105
SAVE THE BAG BAN, YES ON 67, SPONSORED BY ENVIRONMENT CALIFORNIA                                                                                             3027
COMMITTEE TO DEFEAT PROP 23, SPONSORED BY ENVIRONMENT CALIFORNIA                                                                                             3027
Name: committee_name, dtype: int64

Create a subset by filtering contribs using the *complete* committee name of the committee with by far the largest number of contributions.

In [13]:
pocontribs = contribs[contribs['committee_name'] == 'CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIATION TRUTH IN AMERICAN GOVERNMENT FUND; NO ON PROPOSITION 62, YES ON PROPOSITION 66 (NON-PROFIT 501 (C) 5)']

In [14]:
pocontribs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28617 entries, 103 to 28719
Data columns (total 13 columns):
calaccess_committee_id          28617 non-null int64
committee_name                  28617 non-null object
calaccess_filing_id             28617 non-null int64
date_received                   28617 non-null object
contributor_lastname            28617 non-null object
contributor_firstname           28612 non-null object
contributor_city                28617 non-null object
contributor_state               28617 non-null object
contributor_zip                 28617 non-null object
contributor_employer            28612 non-null object
contributor_occupation          28612 non-null object
contributor_is_self_employed    28617 non-null bool
amount                          28617 non-null float64
dtypes: bool(1), float64(1), int64(2), object(9)
memory usage: 2.9+ MB


There were *many* contributions to this committee of $287.08.

In [15]:
pocontribs.describe()

calaccess_committee_id  calaccess_filing_id        amount
count                 28617.0         2.861700e+04  28617.000000
mean                1302403.0         2.013299e+06    291.547605
std                       0.0         4.095908e+02    342.770866
min                 1302403.0         1.982315e+06    287.080000
25%                 1302403.0         2.013304e+06    287.080000
50%                 1302403.0         2.013304e+06    287.080000
75%                 1302403.0         2.013304e+06    287.080000
max                 1302403.0         2.013304e+06  29795.500000

In [16]:
pocontribs['amount'].min()

287.07999999999998

In [17]:
pocontribs[pocontribs['amount'] == pocontribs['amount'].min()]

calaccess_committee_id  \
108                   1302403   
109                   1302403   
110                   1302403   
111                   1302403   
112                   1302403   
113                   1302403   
114                   1302403   
115                   1302403   
116                   1302403   
117                   1302403   
118                   1302403   
119                   1302403   
120                   1302403   
121                   1302403   
122                   1302403   
123                   1302403   
124                   1302403   
125                   1302403   
126                   1302403   
127                   1302403   
128                   1302403   
129                   1302403   
130                   1302403   
131                   1302403   
132                   1302403   
133                   1302403   
134                   1302403   
135                   1302403   
136                   1302403   
137                   1302403   
...                       ...   
28690                 1302403   
28691                 1302403   
28692                 1302403   
28693                 1302403   
28694                 1302403   
28695                 1302403   
28696                 1302403   
28697                 1302403   
28698                 1302403   
28699                 1302403   
28700                 1302403   
28701                 1302403   
28702                 1302403   
28703                 1302403   
28704                 1302403   
28705                 1302403   
28706                 1302403   
28707                 1302403   
28708                 1302403   
28709                 1302403   
28710                 1302403   
28711                 1302403   
28712                 1302403   
28713                 1302403   
28714                 1302403   
28715                 1302403   
28716                 1302403   
28717                 1302403   
28718                 1302403   
28719                 1302403   

                                          committee_name  calaccess_filing_id  \
108    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
109    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
110    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
111    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
112    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
113    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
114    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
115    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
116    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
117    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
118    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
119    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
120    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
121    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
122    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
123    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
124    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
125    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
126    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
127    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
128    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
129    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
130    CALIFORNIA CORRECTIONAL PEACE OFFICERS ASSOCIA...              2013304   
131    CALIFORNI

Create another subset, filtering with other comittee data.

In [20]:
prop = props[props.prop_name == 'PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUTES OF 2014), LARA. ENGLISH LANGUAGE EDUCATION']

In [21]:
prop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4 entries, 48 to 51
Data columns (total 8 columns):
ocd_prop_id               4 non-null object
calaccess_prop_id         4 non-null int64
ccdc_prop_id              4 non-null int64
prop_name                 4 non-null object
ccdc_committee_id         4 non-null int64
calaccess_committee_id    4 non-null int64
committee_name            4 non-null object
committee_position        4 non-null object
dtypes: int64(4), object(4)
memory usage: 288.0+ bytes


In [23]:
propcontribs = pd.merge(prop, contribs, on='calaccess_committee_id')

In [24]:
propcontribs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72 entries, 0 to 71
Data columns (total 20 columns):
ocd_prop_id                     72 non-null object
calaccess_prop_id               72 non-null int64
ccdc_prop_id                    72 non-null int64
prop_name                       72 non-null object
ccdc_committee_id               72 non-null int64
calaccess_committee_id          72 non-null int64
committee_name_x                72 non-null object
committee_position              72 non-null object
committee_name_y                72 non-null object
calaccess_filing_id             72 non-null int64
date_received                   72 non-null object
contributor_lastname            72 non-null object
contributor_firstname           9 non-null object
contributor_city                72 non-null object
contributor_state               72 non-null object
contributor_zip                 72 non-null object
contributor_employer            8 non-null object
contributor_occupation          8 non

In [25]:
propcontribs.amount.sum()

4409450.0

In [26]:
propsupport = propcontribs[propcontribs.committee_position == 'SUPPORT']

In [27]:
propoppose = propcontribs[propcontribs.committee_position == 'OPPOSE']

In [28]:
propsupport.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72 entries, 0 to 71
Data columns (total 20 columns):
ocd_prop_id                     72 non-null object
calaccess_prop_id               72 non-null int64
ccdc_prop_id                    72 non-null int64
prop_name                       72 non-null object
ccdc_committee_id               72 non-null int64
calaccess_committee_id          72 non-null int64
committee_name_x                72 non-null object
committee_position              72 non-null object
committee_name_y                72 non-null object
calaccess_filing_id             72 non-null int64
date_received                   72 non-null object
contributor_lastname            72 non-null object
contributor_firstname           9 non-null object
contributor_city                72 non-null object
contributor_state               72 non-null object
contributor_zip                 72 non-null object
contributor_employer            8 non-null object
contributor_occupation          8 non

In [29]:
propoppose.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 20 columns):
ocd_prop_id                     0 non-null object
calaccess_prop_id               0 non-null int64
ccdc_prop_id                    0 non-null int64
prop_name                       0 non-null object
ccdc_committee_id               0 non-null int64
calaccess_committee_id          0 non-null int64
committee_name_x                0 non-null object
committee_position              0 non-null object
committee_name_y                0 non-null object
calaccess_filing_id             0 non-null int64
date_received                   0 non-null object
contributor_lastname            0 non-null object
contributor_firstname           0 non-null object
contributor_city                0 non-null object
contributor_state               0 non-null object
contributor_zip                 0 non-null object
contributor_employer            0 non-null object
contributor_occupation          0 non-null object
contributor_

In [30]:
propsupport.amount.sum()

4409450.0

In [31]:
propoppose.amount.sum()

0

In [32]:
round(propsupport.amount.sum() / propcontribs.amount.sum(), 2)

1.0

In [33]:
from decimal import Decimal

In [34]:
round(Decimal(propsupport.amount.sum() / propcontribs.amount.sum()), 2)

Decimal('1.00')

In [35]:
propsupport.sort_values('amount', ascending=False).head()

ocd_prop_id  calaccess_prop_id  \
67  ocd-contest/7476a546-1615-41f7-8c7a-f07256c81303            1382395   
22  ocd-contest/7476a546-1615-41f7-8c7a-f07256c81303            1382395   
9   ocd-contest/7476a546-1615-41f7-8c7a-f07256c81303            1382395   
66  ocd-contest/7476a546-1615-41f7-8c7a-f07256c81303            1382395   
70  ocd-contest/7476a546-1615-41f7-8c7a-f07256c81303            1382395   

    ccdc_prop_id                                          prop_name  \
67            77  PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...   
22            77  PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...   
9             77  PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...   
66            77  PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...   
70            77  PROPOSITION 058 - SB 1174 (CHAPTER 753, STATUT...   

    ccdc_committee_id  calaccess_committee_id  \
67                432                 1391170   
22                430                 1386477   
9                 430                 1386477   
66                432                 1391170   
70                432                 1391170   

                                     committee_name_x committee_position  \
67  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...            SUPPORT   
22  YES ON 58, CALIFORNIANS FOR ENGLISH PROFICIENC...            SUPPORT   
9   YES ON 58, CALIFORNIANS FOR ENGLISH PROFICIENC...            SUPPORT   
66  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...            SUPPORT   
70  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...            SUPPORT   

                                     committee_name_y  calaccess_filing_id  \
67  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...              2098635   
22  YES ON 58, CALIFORNIANS FOR ENGLISH PROFICIENC...              2098326   
9   YES ON 58, CALIFORNIANS FOR ENGLISH PROFICIENC...              2083445   
66  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...              2098635   
70  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...              2117213   

   date_received                        contributor_lastname  \
67    2016-10-13                                      STEYER   
22    2016-10-06  CALIFORNIA TEACHERS ASSOCIATION/ISSUES PAC   
9     2016-08-11  CALIFORNIA TEACHERS ASSOCIATION/ISSUES PAC   
66    2016-09-29                          NEXTGEN CALIFORNIA   
70    2016-10-25                                      STEYER   

   contributor_firstname contributor_city contributor_state contributor_zip  \
67                THOMAS    SAN FRANCISCO                CA      94104-4919   
22                   NaN       BURLINGAME                CA           94010   
9                    NaN       BURLINGAME                CA           94010   
66                   NaN    SAN FRANCISCO                CA      94104-4527   
70                THOMAS    SAN FRANCISCO                CA      94104-4919   

   contributor_employer contributor_occupation  contributor_is_self_employed  \
67     FARALLON CAPITAL                FOUNDER                         False   
22                  NaN                    NaN                         False   
9                   NaN                    NaN                         False   
66                  NaN                    NaN                         False   
70     FARALLON CAPITAL                FOUNDER                         False   

       amount  
67  1500000.0  
22   500000.0  
9    500000.0  
66   250000.0  
70   250000.0

In [36]:
propoppose.sort_values('amount', ascending=False).head()

Empty DataFrame
Columns: [ocd_prop_id, calaccess_prop_id, ccdc_prop_id, prop_name, ccdc_committee_id, calaccess_committee_id, committee_name_x, committee_position, committee_name_y, calaccess_filing_id, date_received, contributor_lastname, contributor_firstname, contributor_city, contributor_state, contributor_zip, contributor_employer, contributor_occupation, contributor_is_self_employed, amount]
Index: []

In [37]:
propcontribs.groupby('committee_name_x').amount.sum().reset_index().sort_values('amount', ascending=False)

committee_name_x     amount
1  MILLION VOTER PROJECT ACTION FUND - YES ON 55,...  2110000.0
2  YES ON 58, CALIFORNIANS FOR ENGLISH PROFICIENC...  2009250.0
0  LARA BALLOT MEASURE COMMITTEE; CALIFORNIANS FO...   290200.0

In [38]:
propcontribs.groupby(['contributor_firstname','contributor_lastname','contributor_zip','committee_position']).amount.sum().reset_index().sort_values('amount', ascending=False)

contributor_firstname  \
6                            THOMAS   
0  CAS FOR A 21ST CENTURY ECONOMY -   
3                              LYNN   
4                             PETRA   
2                             LAURA   
1                              KENO   
5                              RYAN   

                                contributor_lastname contributor_zip  \
6                                             STEYER      94104-4919   
0  YES ON 58, A RICARDO LARA BALLOT MEASURE COMMI...           90017   
3                                               AOKI           92841   
4                                            DEJESUS           94112   
2                                               KOHN           92024   
1                                             TESORO           94558   
5                                             BOSSON           95010   

  committee_position     amount  
6            SUPPORT  1750000.0  
0            SUPPORT    49500.0  
3            SUPPORT      200.0  
4            SUPPORT      200.0  
2            SUPPORT      150.0  
1            SUPPORT      100.0  
5            SUPPORT      100.0

In [39]:
props[60:80].prop_name.value_counts()

PROPOSITION 062- DEATH PENALTY. INITIATIVE STATUTE.                                           7
PROPOSITION 064- MARIJUANA LEGALIZATION. INITIATIVE STATUTE.                                  6
PROPOSITION 063- FIREARMS. AMMUNITION SALES. INTIATIVE STATUTE.                               4
PROPOSITION 061- STATE PRESCRIPTION DRUG PURCHASES. PRICING STANDARDS. INITIATIVE STATUTE.    3
Name: prop_name, dtype: int64